File: my-folder/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        output = []
        max_rows = len(matrix) - 1
        max_cols = len(matrix[0]) - 1
        curr_row = 0
        curr_col = 0
        while curr_row <= max_rows and curr_col <= max_cols:
            
            for c in range(curr_col, max_cols + 1):
                output.append(matrix[curr_row][c])
            
            curr_row += 1

            for r in range(curr_row, max_rows + 1):
                output.append(matrix[r][max_cols])

            max_cols -= 1

            if not (curr_row <= max_rows and curr_col <= max_cols):
                break

            for c in range(max_cols, curr_col - 1, -1):
                output.append(matrix[max_rows][c])
            
            max_rows -= 1
            
            for r in range(max_rows, curr_row - 1, -1):
                output.append(matrix[r][curr_col])
            
            curr_col += 1

        return output
        

File: my-folder/0054-spiral-matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int rows = matrix.length;
        int cols = matrix[0].length;
        int top = 0;
        int bot = rows;
        int left = 0;
        int right = cols;
        while(left < right && top < bot){
            for(int idx = left; idx < right; idx++){
                res.add(matrix[top][idx]);
            }
            top++;
            for(int idx = top; idx < bot; idx++){
                res.add(matrix[idx][right - 1]);
            }

            right--;

            if (!(left < right && top < bot)) {
                break;
            }

            for(int idx = right - 1; idx >= left; idx--){
                res.add(matrix[bot - 1][idx]);
            }
            bot--;
            for(int idx = bot - 1; idx >= top; idx--){
                res.add(matrix[idx][left]);
            }
            left++;
        }
    return res;
    }
}

// 1 2 3
// 4 5 6
// 7 8 9
